capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+                                         },
+                                         |pin| {
+                                             pin.make_input();
+                                             Some(pin)
+                                         });
+
+        // setup and return struct
+        SDCard {
+            spi: spi,
+            state: Cell::new(State::Idle),
+            after_state: Cell::new(State::Idle),
+            alarm: alarm,
+            alarm_state: Cell::new(AlarmState::Idle),
+            alarm_count: Cell::new(0),
+            is_initialized: Cell::new(false),
+            card_type: Cell::new(0x00),
+            detect_pin: Cell::new(pin),
+            txbuffer: TakeCell::new(txbuffer),
+            rxbuffer: TakeCell::new(rxbuffer),
+            client: Cell::new(None),
+            client_buffer: TakeCell::empty(),
+            client_offset: Cell::new(0),
+        }
+    }
+
+    fn set_spi_slow_mode(&self) {
+        // set to CPHA=0, CPOL=0, 400 kHZ
+        self.spi.configure(hil::spi::ClockPolarity::IdleLow,
+                           hil::spi::ClockPhase::SampleLeading,
+                           400000);
+    }
+
+    fn set_spi_fast_mode(&self) {
+        // set to CPHA=0, CPOL=0, 4 MHz
+        self.spi.configure(hil::spi::ClockPolarity::IdleLow,
+                           hil::spi::ClockPhase::SampleLeading,
+                           4000000);
+    }
+
+    fn send_command(&self,
+                    cmd: SDCmd,
+                    arg: u32,
+                    mut write_buffer: &'static mut [u8],
+                    mut read_buffer: &'static mut [u8],
+                    recv_len: usize) {
+        // Note: a good default recv_len is 10 bytes. Reading too many bytes
+        //  rarely matters. However, it occasionally matters a lot, so we do
+        //  provided a settable recv_len
+
+        if self.is_initialized() {
+            // device is already initialized
+            self.set_spi_fast_mode();
+        } else {
+            // device is still being initialized
+            self.set_spi_slow_mode();
+        }
+
+        // check that write buffer is long enough
+        if write_buffer.len() < 8 + recv_len {
+            panic!("Write buffer too short to send SD card commands");
+        }
+
+        // send dummy bytes to start
+        write_buffer[0] = 0xFF;
+        write_buffer[1] = 0xFF;
+
+        // command
+        if (0x80 & cmd as u8) != 0x00 {
+            // application-specific command
+            write_buffer[2] = 0x40 | (0x7F & cmd as u8);
+        } else {
+            // normal command
+            write_buffer[2] = 0x40 | cmd as u8;
+        }
+
+        // argument, MSB first
+        write_buffer[3] = ((arg >> 24) & 0xFF) as u8;
+        write_buffer[4] = ((arg >> 16) & 0xFF) as u8;
+        write_buffer[5] = ((arg >> 8) & 0xFF) as u8;
+        write_buffer[6] = ((arg >> 0) & 0xFF) as u8;
+
+        // CRC is ignored except for CMD0 and maybe CMD8
```

Comment:
Ha. So we actually set the CRC to 0x95 as default, which is the CRC for CMD0. Every other command needs a CRC  but it doesn't matter what the value. The established practice appears to be just to use the CMD0 CRC for everything.